proselint/checks/__init__.py

```diff
@@ -1 +1,394 @@
+        )
+
+    def find_ranges(_text: str) -> list[tuple[int, int]]:
+        # TODO: optimize, as it is top3 time-waster (of module-functions)
+        #       this could be a 1d array / LUT
+        s = 0
+        q = pc = ""
+        start = None
+        ranges = []
+        seps = " .,:;-\r\n"
+        quotes = ['"', "“", "”", "'"]
+        for i, c in enumerate(_text + "\n"):
+            if s == 0 and c in quotes and pc in seps:
+                start = i
+                s = 1
+                q = c
+            elif s == 1 and matching(c, q):
+                s = 2
+            elif s == 2:
+                if c in seps:
+                    ranges.append((start + 1, i - 1))
+                    start = None
+                    s = 0
+                else:
+                    s = 1
+            pc = c
+        return ranges
+
+    def position_in_ranges(ranges: list[tuple[int, int]], _position: int) -> bool:
+        return any(start <= _position < end for start, end in ranges)
+
+    return position_in_ranges(find_ranges(text), position)
+
+
+###############################################################################
+# The actual check-sub-functions used by the checks  ##########################
+###############################################################################
+
+
+def consistency_check(
+    text: str,
+    word_pairs: list,
+    err: str,
+    msg: str,
+    offset: int = 0,
+) -> list[ResultCheck]:
+    """Build a consistency checker for the given word_pairs."""
+    results = []
+
+    for w in word_pairs:
+        matches = [
+            list(re.finditer(w[0], text)),
+            list(re.finditer(w[1], text)),
+        ]
+
+        if len(matches[0]) > 0 and len(matches[1]) > 0:
+            idx_minority = len(matches[0]) > len(matches[1])
+
+            results += [
+                (
+                    m.start() + offset,
+                    m.end() + offset,
+                    err,
+                    msg.format(w[not idx_minority], m.group(0)),
+                    w[not idx_minority],
+                )
+                for m in matches[idx_minority]
+            ]
+
+    return results
+
+
+def preferred_forms_check(  # noqa: PLR0913, PLR0917
+    text: str,
+    items: list,
+    err: str,
+    msg: str,
+    ignore_case: bool = True,
+    offset: int = 0,
+) -> list[ResultCheck]:
+    """Build a checker that suggests the preferred form."""
+    flags = re.IGNORECASE if ignore_case else 0
+    regex = r"[\W^]{}[\W$]"  # TODO: replace with enum below?
+
+    return [
+        (
+            m.start() + 1 + offset,
+            m.end() + offset,
+            err,
+            msg.format(item[0], m.group(0).strip()),
+            item[0],
+        )
+        for item in items
+        for r in item[1]
+        for m in re.finditer(regex.format(r), text, flags=flags)
+    ]
+    # TODO: can we speed up str.format() ?
+    #       fast-string? or do padding already in checks
```

Comment:
Format strings are faster than `.format`, if that's what you're wondering. Whether or not that will work for raw strings, however, I am uncertain.

Reply:
i benchmarked these in `benchmark_str_format.py` and `"this is a %s string" % variable` is fastest, but not by much.